CONTRACTS: Dynamic frame condition checking (#7242)
diffblue/cbmc: src/goto-instrument/contracts/doc/developer/contracts-dev-spec-spec-rewriting.md

# Rewriting Declarative Assign and Frees Specification Functions {#contracts-dev-spec-spec-rewriting}

Back to top @ref contracts-dev-spec

@tableofcontents

The front end allows to express parametric assignable sets and freeable sets
declaratively by calling special built-ins from regular C functions,
as described in @ref contracts-assigns and @ref contracts-frees.

Such functions require a rewriting pass in order to turn them into active

> functions that can effectivel populate a write set with new targets.

Reviewer: ```suggestion
functions that can effectively populate a write set with new targets.
```

Author: fixed

## Rewriting Assigns Clause Functions {#contracts-dev-spec-spec-rewriting-assigns}

GOTO functions which specify assignable locations have the following type
signature:

```c
void foo(parameter-list);
```

> They can have been generated from an assigns clause or written by the user.

Reviewer: ```suggestion
They can be generated from an assigns clause or written by the user.
```

Author: fixed

To convert them to active functions, the first step is to inline their body to

> remove all function calls to check the result to be loop free.

Reviewer: Not sure this sentence makes sense.  Perhaps something like this?

To convert them to active functions, the first step is to inline their body and remove all function calls, the result of which should be loop-free.

Author: wow that is a really bad sentence. using your suggestion

...
CALL __CPROVER_contracts_write_set_insert_assignable(
       __write_set_to_fill, N, ptr, size, is_ptr_to_ptr);
```

The function is then instrumented as described in
@ref contracts-dev-spec-dfcc-instrument which adds a second write set parameter,
that allows to check that the function has no unintended side effects.

```c
void foo(
 <parameter-list>,

 // write set to populate with new targets
 __CPROVER_contracts_write_set_ptr_t __write_set_to_fill,

 // write set against which to check itself for unwanted side effects
  __CPROVER_contracts_write_set_ptr_t __write_set_to_check);
```

## Generating Havoc Functions from Assigns Clause Functions {#contracts-dev-spec-spec-rewriting-havoc}

Contract replacement requires havocing the set of targets described in an
assigns clause. Since there can exist dependencies and aliasing between the
different targets, in order to havoc them in parallel, we first need to snapshot
the target locations into a write set data structure, and generate a function
that can havoc these snapshotted locations in a second step.

GOTO functions which specify assignable locations have the following type
signature:

```c
void foo(parameter-list);
```

> They can have been generated from an assigns clause or written by the user.

Reviewer: ```suggestion
They can be generated from an assigns clause or written by the user.
```

Author: fixed

To convert them to havoc functions, the first step is to inline their body to

> remove all function calls to check the result to be loop free.

Reviewer: Same as above.

Author: fixed
